libraries/AP_Mount/AP_Mount_Siyi.cpp

```diff
@@ -0,0 +1,596 @@
+                send_target_rates(rc_target.pitch, rc_target.yaw, rc_target.yaw_is_ef);
+            } else if (get_rc_angle_target(rc_target)) {
+                send_target_angles(rc_target.pitch, rc_target.yaw, rc_target.yaw_is_ef);
+            }
+            break;
+        }
+
+        // point mount to a GPS point given by the mission planner
+        case MAV_MOUNT_MODE_GPS_POINT: {
+            MountTarget angle_target_rad {};
+            if (get_angle_target_to_roi(angle_target_rad)) {
+                send_target_angles(angle_target_rad.pitch, angle_target_rad.yaw, angle_target_rad.yaw_is_ef);
+            }
+            break;
+        }
+
+        case MAV_MOUNT_MODE_HOME_LOCATION: {
+            MountTarget angle_target_rad {};
+            if (get_angle_target_to_home(angle_target_rad)) {
+                send_target_angles(angle_target_rad.pitch, angle_target_rad.yaw, angle_target_rad.yaw_is_ef);
+            }
+            break;
+        }
+
+        case MAV_MOUNT_MODE_SYSID_TARGET:{
+            MountTarget angle_target_rad {};
+            if (get_angle_target_to_sysid(angle_target_rad)) {
+                send_target_angles(angle_target_rad.pitch, angle_target_rad.yaw, angle_target_rad.yaw_is_ef);
+            }
+            break;
+        }
+
+        default:
+            // we do not know this mode so do nothing
+            break;
+    }
+}
+
+// return true if healthy
+bool AP_Mount_Siyi::healthy() const
+{
+    // unhealthy until gimbal has been found and replied with firmware version info
+    if (!_initialised || !_got_firmware_version) {
+        return false;
+    }
+
+    // unhealthy if attitude information NOT received within the last second
+    if (AP_HAL::millis() - _last_current_angle_rad_ms > 1000) {
+        return false;
+    }
+
+    // if we get this far return healthy
+    return true;
+}
+
+// get attitude as a quaternion.  returns true on success
+bool AP_Mount_Siyi::get_attitude_quaternion(Quaternion& att_quat)
+{
+    att_quat.from_euler(_current_angle_rad.x, _current_angle_rad.y, _current_angle_rad.z);
+    return true;
+}
+
+// reading incoming packets from gimbal and confirm they are of the correct format
+// results are held in the _parsed_msg structure
+void AP_Mount_Siyi::read_incoming_packets()
+{
+    // check for bytes on the serial port
+    int16_t nbytes = MIN(_uart->available(), 1024U);
+    if (nbytes <= 0 ) {
+        return;
+    }
+
+    // flag to allow cases below to reset parser state
+    bool reset_parser = false;
+
+    // process bytes received
+    for (int16_t i = 0; i < nbytes; i++) {
+        const uint8_t b = _uart->read();
+        _msg_buff[_msg_buff_len++] = b;
+
+#if AP_MOUNT_SIYI_DEBUG
+        // debug
+        _tot_num_bytes++;
+#endif
+
+        // protect against overly long messages
+        if (_msg_buff_len > AP_MOUNT_SIYI_PACKETLEN_MAX) {
+            reset_parser = true;
+            _parsed_msg.error_count++;
```

Comment:
Why is the error count in `parsed_msg`?

Reply:
I could move error_count outside the sturcture if you feel strongly about it.  This variable holds the number of parsing error.. so while it's not tied to a particular message it is tied to the parsing in general.  The state variable is similar.

Reply:
In the interests of reducing flash I've removed the error_count which was only used for debug purposes.